tools/its.py

```diff
@@ -33,127 +35,159 @@
 
 
 class ImpfterminService():
-    def __init__(self, code: str, plz_impfzentren: list, kontakt: dict, PATH: str):
-        self.code = str(code).upper()
-        self.splitted_code = self.code.split("-")
-
+    def __init__(self, codes: list, plz_impfzentren: list, kontakt: dict, PATH: str):
         self.PATH = PATH
-
-        # PLZ's zu String umwandeln
-        self.plz_impfzentren = sorted([str(plz) for plz in plz_impfzentren])
-        self.plz_termin = None
-
         self.kontakt = kontakt
-        self.authorization = b64encode(bytes(f":{code}", encoding='utf-8')).decode("utf-8")
 
         # Logging einstellen
         self.log = CLogger("impfterminservice")
-        self.log.set_prefix(f"*{self.code[-4:]} | {', '.join(self.plz_impfzentren)}")
+        self.log.set_prefix(', '.join(plz_impfzentren))
 
         # Session erstellen
         self.s = cloudscraper.create_scraper()
         self.s.headers.update({
-            'Authorization': f'Basic {self.authorization}',
             'User-Agent': 'Mozilla/5.0',
         })
 
         # Ausgewähltes Impfzentrum prüfen
-        self.verfuegbare_impfzentren = {}
-        self.domain = None
-        if not self.impfzentren_laden():
-            raise ValueError("Impfzentren laden fehlgeschlagen")
+        while True:
+            try:
+                self.impfzentren = self.impfzentren_laden(plz_impfzentren)
+                break
+            except RuntimeError as exc:
+                self.log.error(str(exc))
+                # Neuer Versuch in nächster Iteration.
+                time.sleep(30)
+
+
+        # Zunächst alle Codes zu allen URLs zuordnen. Aussortiert wird später.
+        self.codes = {
+            url: list(copy.copy(codes))
+            for url in self.impfzentren
+        }
+        self.update_log_prefix()
 
         # Verfügbare Impfstoffe laden
-        self.verfuegbare_qualifikationen: List[Dict] = []
-        while not self.impfstoffe_laden():
-            self.log.warn("Erneuter Versuch in 60 Sekunden")
-            time.sleep(60)
+        self.verfuegbare_qualifikationen = {}
+        for url in self.impfzentren:
+            try:
+                self.verfuegbare_qualifikationen[url] = self.impfstoffe_laden(url)
+            except RuntimeError as exc:
+                # Wissen der verfügbare Impfstoffe wird nicht zwingend benötigt,
+                # also nur ein Warning:
+                self.log.warn(str(exc))
+
+        self.cookies = {}
```

Comment:
```suggestion
        self.cookies = {}
        # OS  
        self.operating_system = platform.system().lower()
```

Reply:
Zu deinem Cookie Problem funktioniert in der Beta bei mir alles. Grundsätzlich muss beachtet werden Headless=invalide Cookies und bei der Request Session dürfen nur der bm_sz und der akavapau Cookie gesetzt werden.

Reply:
Ich hab ja immer noch den Code drin, durch den es beim zweiten und allen weiteren Versuchen ohne Headless und mit 30 Sekunden Wartezeit probiert wird. Das wird aktiviert durch das `manual=True`: https://github.com/iamnotturner/vaccipy/blob/d94725705da761ab8303c89ea2209217087af1ec/tools/its.py#L747

Außerdem:

> bei der Request Session dürfen nur der bm_sz und der akavapau Cookie gesetzt werden.

Sicher? Ich hatte eine Weile alten Code genutzt, bei dem man die Cookies noch manuell eingeben musste. Da hab ich immer alle Cookies eingegeben und das hat funktioniert. Den alten Code kann ich jetzt aber ohne Aufwand nicht mehr nutzen um das gegenzuprüfen, weil ja damals noch nicht das Geburtsdatum übermittelt werden musste.

Jedenfalls hab ich es gerade auch mal probiert, dass nur `bm_sz` und `akavpau_User_allowed` gesetzt werden, komme aber trotzdem nur in die Botprotection.

Reply:
Ja meiner Meinung nach kann man das Headless komplett weglassen, da es ja eh nichts bringt. Aktuell ist es nur sinnvoll, da die User dann direkt einen Fehler bekommen wenn sie nur die Suche starten und ein Selenium Fehler direkt fliegt und nicht erst beim Buchen. 

Funktioniert die Beta bei dir? Kann auch gut sein dass du durch vieles Testen erst mal temporär eine härtere Botprotection hast. Die Cookies sind immer nur einen Request valide und das löschen der Cookies ist glaub ich auch wichtig um eine Historie zu vermeiden. Die Beta funktioniert bei mir mehrmals hintereinander Problemlos und erzeugt valide Cookies.

Reply:
> Aktuell ist es nur sinnvoll, da die User dann direkt einen Fehler bekommen wenn sie nur die Suche starten und ein Selenium Fehler direkt fliegt und nicht erst beim Buchen.

Meinst du damit, dass man eigentlich gar keine Cookies braucht, um eine Reservierung über `rest/suche/impfterminsuche` zu finden?

Reply:
> > Aktuell ist es nur sinnvoll, da die User dann direkt einen Fehler bekommen wenn sie nur die Suche starten und ein Selenium Fehler direkt fliegt und nicht erst beim Buchen.
> 
> Meinst du damit, dass man eigentlich gar keine Cookies braucht, um eine Reservierung über `rest/suche/impfterminsuche` zu finden?

Genau in den neusten Versionen werden ja die Cookies sogar absichtlich vor jedem Request gelöscht. Vermutung ist, dass alte Sessions weniger Termine zugewiesen bekommen. Hatte teilweise das Verhalten, dass das Skript Stunden lief und keinen Termin bekommen hat und manuell waren über 10 min Termine verfügbar. Neustart von Skript hat dann auch geholfen.(Das war vor dem regelmäßigen Cookie löschen)

Habe es mit deinem aktuellen Flow probiert mit dem Code:
```cookie = driver.get_cookie("bm_sz").get("value")
        akavpau = driver.get_cookie("akavpau_User_allowed").get("value")

        cookies ={"bm_sz": cookie,"akavpau_User_allowed": akavpau }```
Werde da auch geblockt.

Reply:
cookies=self.cookies[url] kann sein es liegt daran. Würde die Cookies immer der Session zuweisen.

Reply:
> Genau in den neusten Versionen werden ja die Cookies sogar absichtlich vor jedem Request gelöscht.

Wie kommt es dann, dass ein gefundener Termin für den Impf-Code reserviert ist? Oder ist das nicht mehr so, da jetzt werder Cookies noch Code an `rest/suche/impfterminsuche` übermittelt werden? Oder ist der Termin erst für den Code reserviert, nachdem man über `rest/buchung` in die Botprotection gelaufen ist?

Reply:
Die Reservierung erfolgt serverseitig und ist dann an den Code gebunden. Der Code wird im Header als base64 übermittelt `Authorization: Basic OkZHS......GRFZW`. Cookies spielen keine Rolle bei der Reservierung.

Reply:
Ah stimmt, hab vergessen, dass dieser Header an `rest/suche/impfterminsuche` ebenfalls übermittelt wird.

Reply:
> cookies=self.cookies[url] kann sein es liegt daran.

Ich hab mir das HTTP-Request angeschaut und der `Cookie`-Header wird definitiv gesendet:

```
POST /rest/smspin/anforderung HTTP/1.1
Host: 001-iz.impfterminservice.de
User-Agent: Mozilla/5.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Language: en-US,en;q=0.9
Accept-Encoding: gzip, deflate
Cookie: akavpau_User_allowed=<redacted>; bm_sz=<redacted>
Content-Length: 128
Content-Type: application/json

{"plz": "70376", "email": "<redacted>", "phone": "<redacted>", "birthday": "1982-03-02", "einzeltermin": false}
```

Reply:
Ja die werden auch benötigt. Hauptsache die Response ist nicht 429.

Reply:
> Werde da auch geblockt.

@sinun98 Ist jetzt durch folgende Änderung gefixt:

Ich rufe immer `self.s.cookies.clear()` vor jedem Request auf.

Ich wusste nicht, dass der Scraper die Cookies aus den Responses ebenfalls speichert. In meinem Test ist das nicht aufgefallen, da ich zum Abfangen der HTTP-Requests eine anderen Host genommen habe und der Scraper die Cookies (aus gutem Grund) nur an die richtigen Hosts sendet.

Reply:
Ja funktioniert jetzt bei mir auch. Der Scraper ist ja eigentlich für Cloudflare gedacht. Headers kann man z.B. auch nicht direkt mitgeben. 
Noch besser funktioniert es bei mir btw wenn man beim ersten Versuch Cookies zu generieren nicht Headless startet. Dann werden zumindest bei mir die Cookies von selbst erzeugt und ich muss gar nicht manuell eingreifen.

Reply:
> Headers kann man z.B. auch nicht direkt mitgeben.

Doch, mache ich an drei Stellen, jeweils mit dem `headers=` Parameter.